### ISSUE TITLE
MGLVectorStyleLayer.sourceLayerIdentifier is never optional

### ISSUE THREAD

**Issue by incanus** [mapbox/mapbox-gl-native#6780]
Upstream spec issue: https://github.com/mapbox/mapbox-gl-style-spec/issues/547

Vector layer styling doesn't work if the `sourceLayerIdentifier` isn't specified, even if there is only one data layer in the source and/or it is named just like the source identifier and/or layer identifier. 

While we're in here, we could consider a few enhancements: 
- Auto-determining single-data-layer source layer names
- A better property name such as `sourceDataLayerIdentifier` or `sourceLayerName` to emphasize that sources have _data_ layers that tie to _style_ layers
- KVO-aware source layer querying a là style layer querying in https://github.com/mapbox/mapbox-gl-native/pull/6097

/cc @1ec5 @boundsj

**Comment by incanus** [comment]
Some background on iOS: https://github.com/mapbox/mapbox-gl-native/pull/6269#issuecomment-245178503

**Comment by 1ec5** [comment]
https://github.com/mapbox/mapbox-gl-style-spec/issues/547#issuecomment-255287115 points out that the source layer is never specified for GeoJSON sources. The documentation could clarify the fact that the source layer is required for multilayered vector sources. We could also bring back the `sourceLayerIdentifier` initializer parameter but leave it `nullable` for the vector source case.

> Auto-determining single-data-layer source layer names

If after mapbox/mapbox-gl-style-spec#547 the style specification continues to carve out an exception for single-layer sources, but mbgl doesn’t automatically use the sole layer in a source, please open a separate issue.

**Comment by ericrwolfe** [comment]
Was just scratching my head as to why a layer wasn't appearing. 

We ought to consider adding an assertion if the source layer referenced by the style layer isn't found.

**Comment by 1ec5** [comment]
> Auto-determining single-data-layer source layer names
> We ought to consider adding an assertion if the source layer referenced by the style layer isn't found.

Defaulting to the sole source layer should happen at the mbgl level. The SDK currently can neither do that nor assert that a source layer identifier is valid, because mbgl doesn’t expose any information about what constitutes a vector source (#7255), especially not a source baked into the style and obtained through `-sourceWithIdentifier:` (#6584).

**Comment by 1ec5** [comment]
Moving off a milestone, because further action depends on #7255.

**Comment by 1ec5** [comment]
An interim fix could look like this:

* [ ] Document the fact that a style layer identifier is required when working with MGLVectorSource
* [ ] When adding an MGLVectorStyleLayer to an MGLMapView, if the source refers to an MGLVectorSource but there’s no style layer identifier, raise an exception
